### PR TITLE
Rich schema schema txn's

### PR DESCRIPTION
**Pull request by burdettadam**
This PR has code for rich schema, schema transactions.

### CODE REVIEW

File: indy_common/test/types/test_get_rs_schema_schema.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
+from collections import OrderedDict
+
+from indy_common.constants import RS_META_VERSION, RS_META_NAME, RS_META_TYPE, RS_SCHEMA_FROM, TXN_TYPE, RS_META
+from indy_common.types import ClientGetRsSchemaOperation, RsSchemaMetaField
+from plenum.common.messages.fields import ConstantField, IdentifierField, VersionField, LimitedLengthStringField
+
+EXPECTED_ORDERED_FIELDS_SCHEMA = OrderedDict([
```
Comment (ashcherbakov) [inline indy_common/test/types/test_get_rs_schema_schema.py]:
```suggestion
EXPECTED_ORDERED_FIELDS_SCHEMA_META = OrderedDict([
```

File: indy_common/test/types/test_rs_schema_schema.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+def test_meta_has_expected_validators_s():
+    schema = dict(RsSchemaMetaField.schema)
+    for field, validator in EXPECTED_SCHEMA_META_FIELDS.items():
+        assert isinstance(schema[field], validator)
+
+
+EXPECTED_RS_SCHEMA_DATA_FIELDS = OrderedDict([
+    (RS_JSON_LD_TYPE, LimitedLengthStringField),
+    (RS_JSON_LD_ID, LimitedLengthStringField),
```
Comment (ashcherbakov) [inline indy_common/test/types/test_rs_schema_schema.py]: Should we expect `RS_JSON_LD_CONTEXT ` (`@context`) here as well?

File: indy_common/types.py [hunk 3]
```diff
@@ -142,6 +143,54 @@ class GetContextField(MessageValidator):
...
+    )
+    _base_types = None
+
+    def __init__(self, max_size=None, **kwargs):
+        if max_size is not None:
+            if not isinstance(max_size, int):
+                raise PlenumTypeError('max_size', max_size, int)
+            if not max_size > 0:
+                raise PlenumValueError('max_size', max_size, '> 0')
+        self.max_size = max_size
```
Comment (ashcherbakov) [inline indy_common/types.py]: Where `max_size` is used? Do we expect to add `_specific_validation` here in future?

File: indy_node/server/request_handlers/domain_req_handlers/rs_schema_handler.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    def update_state(self, txn, prev_result, request, is_committed=False) -> None:
+        self._validate_txn_type(txn)
+        path, value_bytes = RsSchemaHandler.prepare_state(txn)
+        self.state.set(path, value_bytes)
+
+    @staticmethod
+    def prepare_state(txn, path_only=False):
+        payload = txn["txn"]
+        meta, data = payload[RS_META], payload[RS_DATA]
+        did_author = txn["txn"]["metadata"].get("from", None)  # why not txn.identifier
```
Comment (ashcherbakov) [inline indy_node/server/request_handlers/domain_req_handlers/rs_schema_handler.py]: `get_from` helper method can be used here instead

File: indy_node/server/request_handlers/domain_req_handlers/rs_schema_handler.py [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+        _id = RsSchemaHandler.make_state_path(did_author, meta[RS_META_NAME], meta[RS_META_VERSION])
+        if path_only:
+            return _id
+        value = {
+            RS_META: meta,
+            RS_DATA: data
+        }
+        seq_no = txn["txnMetadata"].get("reqSignature", None)
```
Comment (ashcherbakov) [inline indy_node/server/request_handlers/domain_req_handlers/rs_schema_handler.py]: `get_seq_no` can be used here (`reqSignature` looks like a typo).

File: indy_node/server/request_handlers/domain_req_handlers/rs_schema_handler.py [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+        txn_time = txn["txnMetadata"].get("txnTime", None)
```
Comment (ashcherbakov) [inline indy_node/server/request_handlers/domain_req_handlers/rs_schema_handler.py]: `get_txn_time` can be used here

File: indy_node/test/request_handlers/conftest.py [hunk 7]
```diff
@@ -43,6 +49,60 @@ def schema_request():
...
 
 
+@pytest.fixture(scope="module")
+def rs_schema_handler(db_manager, write_auth_req_validator):
+    return RsSchemaHandler(db_manager, write_auth_req_validator)
+
+
+@pytest.fixture(scope="function")
+def rs_schema_request():
+    authors_did, name, version, type = "2hoqvcwupRTUNkXn6ArYzs", randomString(), "1.1", MARKER_RS_SCHEMA
```
Comment (ashcherbakov) [inline indy_node/test/request_handlers/conftest.py]: Since `type` is a language key word, it's better to use `_type` or something similar.

File: indy_node/test/request_handlers/conftest.py [hunk 8]
```diff
@@ -43,6 +49,60 @@ def schema_request():
...
+    _id = authors_did + ':' + type + ':' + name + ':' + version
+    return Request(identifier=authors_did,
+                   reqId=random.randint(1, 10000000000000000000),
+                   signature="sig",
+                   protocolVersion=2,
+                   operation={
+                       'type': SET_RS_SCHEMA,
+                       RS_META: {
+                           RS_META_ID: _id,
```
Comment (ashcherbakov) [inline indy_node/test/request_handlers/conftest.py]: Are `RS_META_ID` and `RS_META_TYPE` mandatory fields or not?

File: docs/source/requests.md [hunk 9]
```diff
@@ -2003,6 +2008,195 @@ So, if the Context needs to be evolved, a new Context with a new version or name
...
+        - `@type`: The value of this property must be (or map to, via a context object) a URI.
+        - `@context`: If present, the value of this property must be a context object or a URI which can be dereferenced to obtain a context object.
+        
+- `meta` (dict)
+
+    Dictionary with Rich Schema's metadata
+
+    - `name` (string): schema's name
+    - `version` (string): schema's version
+    - `type` (string): "sch"
```
[the remark this answers is not in the document]
Comment (burdettadam) [reply docs/source/requests.md]: yes, as far as I know, "sch" is the only supported type for rich schema schema. I do not see any use for it now but it does allow possible flexibility in the future. All information I have can be found in this rfc https://github.com/hyperledger/indy-hipe/tree/490345b9ca3f292dd66cce226bd15bec959c3735/text/0149-rich-schema-schema#set_rich_schema

File: docs/source/requests.md [hunk 10]
```diff
@@ -108,7 +111,7 @@ Each Request (both write and read) is a JSON with a number of common metadata fi
             - GET_REVOC_REG = "116"
             - GET_REVOC_REG_DELTA = "117"  
             - GET_AUTH_RULE = "121"      
-            - GET_CONTEXT = "300"
+            - GET_RICH_SCHEMA = "301"
```
[the remark this answers is not in the document]
Comment (burdettadam) [reply docs/source/requests.md]: I agree, good catch. I will add that back in.

File: docs/source/requests.md [hunk 11]
```diff
@@ -2003,6 +2008,195 @@ So, if the Context needs to be evolved, a new Context with a new version or name
...
+        - `@context`: If present, the value of this property must be a context object or a URI which can be dereferenced to obtain a context object.
+        
+- `meta` (dict)
+
+    Dictionary with Rich Schema's metadata
+
+    - `name` (string): schema's name
+    - `version` (string): schema's version
+    - `type` (string): "sch"
+    - `tag` (string, optional):
```
[the remark this answers is not in the document]
Comment (burdettadam) [reply docs/source/requests.md]: I agree. I will update the markdown.

File: docs/source/transactions.md [hunk 12]
```diff
@@ -735,6 +737,197 @@ If the Context needs to be evolved, a new Context with a new version or new name
 }
 ```
 
+
+#### SET_RICH_SCHEMA
```
[the remark this answers is not in the document]
Comment (burdettadam) [reply docs/source/transactions.md]: I will update.

File: indy_common/constants.py [hunk 13]
```diff
@@ -10,7 +10,26 @@
...
+RS_META_NAME = 'name'
+RS_META_TYPE = 'type'
+RS_META_VERSION = 'version'
+RS_META_ID = 'id'
+RS_DATA = 'data'
+RS_JSON_LD_ID = '@id'
+RS_JSON_LD_TYPE = '@type'
+RS_JSON_LD_CONTEXT = '@context'
+
+# RS_SCHEMA
```
[the remark this answers is not in the document]
Comment (burdettadam) [reply indy_common/constants.py]: I don't remember. I think I wanted a better name than `rich schema schema` so I used `rs_schema` with `RICH SCHEMA` as a general comment. I will remove un-used constants and try to reuse others to clean this section up.

File: indy_common/constants.py [hunk 14]
```diff
@@ -10,7 +10,26 @@
...
+RS_SCHEMA_CONTEXT = "@context"
+RS_SCHEMA_TYPE = "@type"
+RS_SCHEMA_DOCUMENT = "properties"
+RS_SCHEMA_FROM = "dest"
```
[the remark this answers is not in the document]
Comment (burdettadam) [reply indy_common/constants.py]: I strongly agree. I will change `dest` to `from`.

File: indy_common/test/types/test_get_rs_schema_schema.py [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+    (RS_META_TYPE, ConstantField),
```
[the remark this answers is not in the document]
Comment (burdettadam) [reply indy_common/test/types/test_get_rs_schema_schema.py]: I understand. I will update the tests to use bare strings instead of constants.

File: indy_common/test/types/test_rs_schema_schema.py (hunk 2 above)
Comment (burdettadam) [reply indy_common/test/types/test_rs_schema_schema.py]: No. `@context` is optional. I hope the RFC will be updated to better reflect this. https://github.com/hyperledger/indy-hipe/blob/master/text/0149-rich-schema-schema/README.md#example-schema .
The current example in the RFC is missing the required `@id` and `@type` while including optional `@context`.

File: indy_common/types.py (hunk 3 above)
Comment (burdettadam) [reply indy_common/types.py]: `max_size` is not used. I am not sure what `_specific_validation` is, but I did not plan to use it in the future.
I copied this code from context, for the size check, I will remove `self.max_size = max_size`, thanks.

File: indy_node/server/request_handlers/domain_req_handlers/rs_schema_handler.py (hunk 5 above)
Comment (burdettadam) [reply indy_node/server/request_handlers/domain_req_handlers/rs_schema_handler.py]: Good catch! `reqSignature` is a bug. I will change that.

File: indy_node/server/request_handlers/domain_req_handlers/rs_schema_handler.py (hunk 4 above)
Comment (burdettadam) [reply indy_node/server/request_handlers/domain_req_handlers/rs_schema_handler.py]: My original reasoning for not using `get_from` was to keep dependencies down and explicit readability. I still favor `txn["txn"]["metadata"].get("from", None)` over `get_from`, do I need to change it?

File: indy_node/test/request_handlers/conftest.py (hunk 7 above)
Comment (burdettadam) [reply indy_node/test/request_handlers/conftest.py]: I will update `type` to `_type`.

File: indy_node/test/request_handlers/conftest.py (hunk 8 above)
Comment (burdettadam) [reply indy_node/test/request_handlers/conftest.py]: `RS_META_ID` is not mandatory I will remove it. `RS_META_TYPE` is mandatory and should equal `RS_SCHEMA_META_TYPE`, I will correct that.

**Comment by skhoroshavin** [comment]
(ci) test this please

File: indy_common/test/types/test_get_rs_schema_schema.py [hunk 16]
```diff
@@ -0,0 +1,39 @@
+from collections import OrderedDict
+from indy_common.types import ClientGetRsSchemaOperation, RsSchemaMetaField
+from plenum.common.messages.fields import ConstantField, IdentifierField, VersionField, LimitedLengthStringField
+
+
+EXPECTED_ORDERED_FIELDS_SCHEMA = OrderedDict([
+    ("type", ConstantField),
```
Comment (anikitinDSR) [inline indy_common/test/types/test_get_rs_schema_schema.py]: I think it would be rather to use constants instead of strings. For example, TYPE constant instead of "type" string.

File: indy_common/test/types/test_rs_schema_schema.py [hunk 17]
```diff
@@ -0,0 +1,65 @@
+from collections import OrderedDict
+
+from indy_common.types import RsSchemaMetaField, ClientSetRsSchemaOperation, SetRsSchemaDataField, RsSchemaField
+from plenum.common.messages.fields import ConstantField, VersionField, LimitedLengthStringField
+
+EXPECTED_SCHEMA_META_FIELDS = OrderedDict([
+    ("name", LimitedLengthStringField),
```
Comment (anikitinDSR) [inline indy_common/test/types/test_rs_schema_schema.py]: The same situation as for `test_get_rs_schema_schema.py`

File: indy_common/types.py [hunk 18]
```diff
@@ -142,6 +143,53 @@ class GetContextField(MessageValidator):
...
+    schema = (
+        (RS_JSON_LD_TYPE, LimitedLengthStringField(max_length=NAME_FIELD_LIMIT)),
+        (RS_JSON_LD_ID, LimitedLengthStringField(max_length=NAME_FIELD_LIMIT)),
+    )
+    _base_types = None
+
+    def __init__(self, max_size=None, **kwargs):
+        if max_size is not None:
+            if not isinstance(max_size, int):
+                raise PlenumTypeError('max_size', max_size, int)
```
Comment (anikitinDSR) [inline indy_common/types.py]: Are there tests for this exception raises?

File: indy_node/server/request_handlers/domain_req_handlers/rs_schema_handler.py [hunk 19]
```diff
@@ -0,0 +1,89 @@
...
+        return path.decode()
+
+    def update_state(self, txn, prev_result, request, is_committed=False) -> None:
+        self._validate_txn_type(txn)
+        path, value_bytes = RsSchemaHandler.prepare_state(txn)
+        self.state.set(path, value_bytes)
+
+    @staticmethod
+    def prepare_state(txn, path_only=False):
+        payload = txn["txn"]
```
Comment (anikitinDSR) [inline indy_node/server/request_handlers/domain_req_handlers/rs_schema_handler.py]: There are a lot of helpers methods in `plenum/common/txn_util.py` module. I think it would be rather to use they instead of using txn as a dict.

File: indy_node/test/request_handlers/conftest.py [hunk 20]
```diff
@@ -36,7 +42,7 @@ def schema_request():
     return Request(identifier=randomString(),
                    reqId=5,
                    signature="sig",
-                   operation={'type': SCHEMA,
+                   operation={'type': "101",
```
Comment (anikitinDSR) [inline indy_node/test/request_handlers/conftest.py]: Why did you change constant's using to string? If SCHEMA constant will be change by some reason then it will to hard to change all of strings "101" to another instead of changing constant's value.

File: indy_node/test/request_handlers/conftest.py [hunk 21]
```diff
@@ -45,6 +51,83 @@ def schema_request():
...
+def rs_schema_handler(db_manager, write_auth_req_validator):
+    return RsSchemaHandler(db_manager, write_auth_req_validator)
+
+
+@pytest.fixture(scope="function")
+def rs_schema_request():
+    authors_did, name, version, _type = "2hoqvcwupRTUNkXn6ArYzs", randomString(), "1.1", "8"
+    _id = authors_did + ':' + _type + ':' + name + ':' + version
+    return Request(identifier=authors_did,
+                   reqId=random.randint(1, 10000000000000000000),
```
Comment (anikitinDSR) [inline indy_node/test/request_handlers/conftest.py]: You already created a function for reqId generation in ` indy_node/test/api/helper.py`. I think it can be used here.

File: indy_node/test/rs_schema/test_rs_send_schema.py [hunk 22]
```diff
@@ -0,0 +1,175 @@
...
+from indy_common.types import SetRsSchemaDataField
+from indy_node.test.api.helper import sdk_write_rs_schema_and_check, build_rs_schema_request
+from plenum.common.exceptions import RequestRejectedException
+
+
+def test_send_rs_schema_multiple_attrib(looper, sdk_pool_handle, sdk_wallet_endorser):
+    _, identifier = sdk_wallet_endorser
+    authors_did, name, version, type = identifier, "ISO18023_Drivers_License", "1.1", "8"
+    _id = identifier + ':' + type + ':' + name + ':' + version
+    schema = {
```
Comment (anikitinDSR) [inline indy_node/test/rs_schema/test_rs_send_schema.py]: I think, that this definition can be moved to fixture of template or something like that for reusing in several places.

File: indy_common/test/types/test_get_rs_schema_schema.py (hunk 16 above)
Comment (burdettadam) [reply indy_common/test/types/test_get_rs_schema_schema.py]: I originally had constants but as @ashcherbakov pointed out to me, the tests should catch accidental API changes. I prefer tests not having dependances on constants. If you strongly disagree I change tests to use constants.

File: indy_common/test/types/test_rs_schema_schema.py (hunk 17 above)
Comment (burdettadam) [reply indy_common/test/types/test_rs_schema_schema.py]: see the response in `test_get_rs_schema_schema.py` for the same situation.

File: indy_node/server/request_handlers/domain_req_handlers/rs_schema_handler.py (hunk 19 above)
Comment (burdettadam) [reply indy_node/server/request_handlers/domain_req_handlers/rs_schema_handler.py]: I like to keep dependencies down and believe this is more readable code. My intuition also leads me to believe this is more performant code. As I remember, I do not do this in another place, so using a helper function to access a dictionary feels un-natural.

File: indy_node/test/request_handlers/conftest.py (hunk 20 above)
Comment (burdettadam) [reply indy_node/test/request_handlers/conftest.py]: changing SCHEMA constant is an API breaking change. I think the tests should catch that. I think the benefit of catching accidental API changes in testing is worth the effort required to update test during an API change. 
I can change that back if it goes against the project design.

File: indy_common/types.py (hunk 18 above)
Comment (burdettadam) [reply indy_common/types.py]: There are not any tests for this error. I will add one.

File: indy_node/test/request_handlers/conftest.py (hunk 21 above)
Comment (burdettadam) [reply indy_node/test/request_handlers/conftest.py]: I agree I will add code to use the generator from helper.py

File: indy_node/test/rs_schema/test_rs_send_schema.py (hunk 22 above)
Comment (burdettadam) [reply indy_node/test/rs_schema/test_rs_send_schema.py]: I will add a `templates.py` for reusing the schema structure.

File: indy_node/server/request_handlers/domain_req_handlers/rs_schema_handler.py (hunk 19 above)
Comment (ashcherbakov) [reply indy_node/server/request_handlers/domain_req_handlers/rs_schema_handler.py]: Using helper function is just more error-prone and readable, as you don't expose internals of how txn looks like. 
Also using helper functions may help a lot in case of changing some constants due to versioning.
